### ISSUE TITLE
YOLOv11 to ONNX (with NMS and dynamic batch)

### ISSUE THREAD

**Issue by Egorundel** [ultralytics/ultralytics#18172]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

How to export YOLOv11.pt (trained on a custom dataset with images 512x512) with dynamic batch and NMS (ONNX model)?

**Comment by Egorundel** [comment]
I have this script, is it normal?

```python
import torch 
import torchvision

import onnx
import onnx_graphsurgeon

import numpy as np 
import onnxruntime as ort 

from collections import OrderedDict

# from utils import yolo_insert_nms

def yolo_insert_nms(path, score_threshold=0.25, iou_threshold=0.7, max_output_boxes=100, simplify=False):
    '''
    http://www.xavierdupre.fr/app/onnxcustom/helpsphinx/api/onnxops/onnx__EfficientNMS_TRT.html
    https://huggingface.co/spaces/muttalib1326/Punjabi_Character_Detection/blob/3dd1e17054c64e5f6b2254278f96cfa2bf418cd4/utils/add_nms.py
    '''
    onnx_model = onnx.load(path)

    if simplify:
        from onnxsim import simplify
        onnx_model, _ = simplify(onnx_model,  overwrite_input_shapes={'images': [1, 3, 512, 512]})

    graph = onnx_graphsurgeon.import_onnx(onnx_model)
    graph.toposort()
    graph.fold_constants()
    graph.cleanup()

    topk = max_output_boxes
    attrs = OrderedDict(plugin_version='1',
                        background_class=-1,
                        max_output_boxes=topk,
                        score_threshold=score_threshold,
                        iou_threshold=iou_threshold,
                        score_activation=False,
                        box_coding=0, )

    outputs = [onnx_graphsurgeon.Variable('num_dets', np.int32, [-1, 1]),
               onnx_graphsurgeon.Variable('det_boxes', np.float32, [-1, topk, 4]),
               onnx_graphsurgeon.Variable('det_scores', np.float32, [-1, topk]),
               onnx_graphsurgeon.Variable('det_classes', np.int32, [-1, topk])]

    graph.layer(op='EfficientNMS_TRT', 
                name="batched_nms", 
                inputs=[graph.outputs[0], 
                        graph.outputs[1]], 
                outputs=outputs, 
                attrs=attrs, )

    graph.outputs = outputs
    graph.cleanup().toposort()

    onnx.save(onnx_graphsurgeon.export_onnx(graph), f'yolo_w_nms.onnx')

class YOLO11(torch.nn.Module):
    def __init__(self, name) -> None:
        super().__init__()
        from ultralytics import YOLO
        # Load a model
        # build a new model from scratch
        # model = YOLO(f'{name}.yaml')  

        # load a pretrained model (recommended for training)
        model = YOLO("./yolo11m.pt")
        self.model = model.model

    def forward(self, x):
        '''https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/tasks.py#L216
        '''
        pred: torch.Tensor = self.model(x)[0] # n 84 8400,
        pred = pred.permute(0, 2, 1)
        boxes, scores = pred.split([4, 80], dim=-1)
        boxes = torchvision.ops.box_convert(boxes, in_fmt='cxcywh', out_fmt='xyxy')

        return boxes, scores



def export_onnx(name='./best.pt'):
    '''export onnx
    '''
    m = YOLO11(name)

    x = torch.rand(1, 3, 512, 512)
    dynamic_axes = {
        'images': {0: '-1'}
    }
    torch.onnx.export(m, x, f'{name}.onnx', 
                      input_names=['images'], 
                      output_names=['boxes', 'scores'], 
                      opset_version=13, 
                      dynamic_axes=dynamic_axes)

    data = np.random.rand(1, 3, 512, 512).astype(np.float32)
    sess = ort.InferenceSession(f'{name}.onnx')
    _ = sess.run(output_names=None, input_feed={'images': data})
    
    import onnx 
    import onnxslim
    model_onnx = onnx.load(f'{name}.onnx')
    model_onnx = onnxslim.slim(model_onnx)
    onnx.save(model_onnx, f'{name}.onnx')


if __name__ == '__main__':

    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--name', type=str, default='yolo11n_tuned')
    parser.add_argument('--score_threshold', type=float, default=0.25)
    parser.add_argument('--iou_threshold', type=float, default=0.65)
    parser.add_argument('--max_output_boxes', type=int, default=100)
    args = parser.parse_args()

    export_onnx(name=args.name)
    
    yolo_insert_nms(path=f'{args.name}.onnx', 
                    score_threshold=args.score_threshold, 
                    iou_threshold=args.iou_threshold, 
                    max_output_boxes=args.max_output_boxes, )
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Egorundel, thank you for your interest in Ultralytics 🚀! We recommend checking out our [Docs](https://docs.ultralytics.com) for guidance, where you can explore [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common questions might already be addressed there!

If your query pertains to a specific 🐛 **Bug Report**, please share a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug effectively.

For custom training/export ❓ **Questions**, such as your ONNX export query, please provide as much detail as possible (e.g., code snippets, command usage, or logs). Ensure your environment is set up properly following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

### ONNX Export Reference
To assist you further, check the [Export Documentation](https://docs.ultralytics.com/modes/export/) for tips on exporting YOLO models, including support for formats like ONNX.

### Upgrade
To confirm you’re working with the latest features and fixes, please upgrade to the newest `ultralytics` version with all dependencies in a **Python>=3.8** and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) environment:

```bash
pip install -U ultralytics
```

### Verified Environments
For convenient setups, explore our verified environments for YOLO:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM ([Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/))
- **Amazon** Deep Learning AMI ([Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))
- **Docker** ([Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

### Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>  
If this badge is green, [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are currently passing, confirming the stable operation of YOLO models across macOS, Windows, and Ubuntu.

This is an automated response 😊. One of our Ultralytics engineers will review your question and provide additional support soon. In the meantime, feel free to explore the resources above and engage with our community: [Discord](https://discord.com/invite/ultralytics), [Discourse](https://community.ultralytics.com), or [Subreddit](https://reddit.com/r/Ultralytics).

**Comment by Egorundel** [comment]
Oh, I've already created a working code for exporting :)

```python
import argparse
import sys
from io import BytesIO
from pathlib import Path

import onnx
import torch
from ultralytics import YOLO

FILE = Path(__file__).resolve()
ROOT = FILE.parents[1]
if str(ROOT) not in sys.path:
    sys.path.append(str(ROOT.resolve()))

from models.common import PostDetect, optim  # noqa: E402

try:
    import onnxsim
except ImportError:
    onnxsim = None


def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument('-w',
                        '--weights',
                        type=str,
                        required=True,
                        help='PyTorch yolov8 weights')
    parser.add_argument('--iou-thres',
                        type=float,
                        default=0.7,
                        help='IOU threshoud for NMS plugin')
    parser.add_argument('--conf-thres',
                        type=float,
                        default=0.25,
                        help='CONF threshoud for NMS plugin')
    parser.add_argument('--topk',
                        type=int,
                        default=100,
                        help='Max number of detection bboxes')
    parser.add_argument('--opset',
                        type=int,
                        default=11,
                        help='ONNX opset version')
    parser.add_argument('--sim',
                        action='store_true',
                        help='simplify onnx model')
    parser.add_argument('--dynamic',
                        type=str,
                        default='batch',
                        help='Dynamic axes for onnx export')
    parser.add_argument('--device',
                        type=str,
                        default='cpu',
                        help='Export ONNX device')
    args = parser.parse_args()
    assert args.dynamic in ['batch', 'all']
    PostDetect.conf_thres = args.conf_thres
    PostDetect.iou_thres = args.iou_thres
    PostDetect.topk = args.topk
    return args


def main(args):
    dynamic_axes = {
        'num_dets': {
            0: 'batch'
        },
        'bboxes': {
            0: 'batch'
        },
        'scores': {
            0: 'batch'
        },
        'labels': {
            0: 'batch'
        }
    }
    if args.dynamic == 'batch':
        dynamic_axes.update({'images': {0: 'batch'}})
    else:
        dynamic_axes.update({'images': {0: 'batch', 2: 'height', 3: 'width'}})

    YOLOv8 = YOLO(args.weights)
    model = YOLOv8.model.fuse().eval()
    for m in model.modules():
        optim(m)
        m.to(args.device)
    model.to(args.device)
    # fixed input shape [1, 3, 640, 640]
    fake_input = torch.randn(1, 3, 512, 512).to(args.device)
    for _ in range(2):
        model(fake_input)
    save_path = args.weights.replace('.pt', '.onnx')
    print('\nStarting ONNX export with onnx %s...' % onnx.__version__)
    with BytesIO() as f:
        torch.onnx.export(
            model,
            fake_input,
            f,
            opset_version=args.opset,
            input_names=['images'],
            output_names=['num_dets', 'bboxes', 'scores', 'labels'],
            dynamic_axes=dynamic_axes)
        f.seek(0)
        onnx_model = onnx.load(f)
    onnx.checker.check_model(onnx_model)
    shapes = [
        'batch', 1, 'batch', args.topk, 4, 'batch', args.topk, 'batch',
        args.topk
    ]
    for i in onnx_model.graph.output:
        for j in i.type.tensor_type.shape.dim:
            j.dim_param = str(shapes.pop(0))
    if args.sim:
        try:
            onnx_model, check = onnxsim.simplify(onnx_model)
            assert check, 'assert check failed'
        except Exception as e:
            print(f'Simplifier failure: {e}')
    onnx.save(onnx_model, save_path)
    print(f'ONNX export success, saved as {save_path}')


if __name__ == '__main__':
    main(parse_args())
```

**Comment by glenn-jocher** [comment]
Thank you for sharing your implementation! It looks like you have developed a comprehensive script for exporting YOLO models to ONNX with dynamic axes and NMS. If you'd like to compare or enhance your approach, you can also refer to the built-in `export` method in the Ultralytics YOLO framework, which supports ONNX export with features like `dynamic=True` and `nms=True`. Details are available in the [Export Documentation](https://docs.ultralytics.com/modes/export/). Great work, and feel free to reach out if you have further questions!

**Comment by visionNoob** [comment]
@glenn-jocher 
As far as I know, the ```nms=True``` is not supported when ```format==ONNX``` (TensorRT too).
https://github.com/ultralytics/ultralytics/blob/f932a611abf0ed0e359deecbb82fa8d361e23cfe/ultralytics/engine/exporter.py#L101-L121

**Comment by glenn-jocher** [comment]
@visionNoob thank you for the comment! Actually, `nms=True` is supported for ONNX export in the latest Ultralytics version. Ensure you're using the updated `ultralytics` package and refer to the [export documentation](https://docs.ultralytics.com/modes/export/) for details. If you encounter issues, feel free to share specifics for further assistance.

**Comment by visionNoob** [comment]
Thank you, @glenn-jocher for your response and guidance! I appreciate the clarification regarding `nms=True` support for ONNX export.

However, I am currently (today is `Jan 02 2025`) using the latest version (8.3.56), and I’ve gone through the documentation and the relevant code in detail. Despite this, I haven’t been able to find any reference to `nms=True` for ONNX export. It’s possible that I may have overlooked something, so I would greatly appreciate it if you could point me to the specific section of the documentation or code that addresses this feature.

Thank you again for your support and continuously improving the Ultralytics!
![image](https://github.com/user-attachments/assets/256b5545-eecf-4239-8316-035de5ebd898)
![image](https://github.com/user-attachments/assets/57e03096-5f2b-484f-80bb-cf769e2e59f6)

**Comment by glenn-jocher** [comment]
Thank you for your detailed follow-up! You are correct that currently, `nms=True` is not supported for ONNX export in the latest version (8.3.56) of Ultralytics. While `nms` functionality is available for certain formats, it is explicitly disabled for ONNX and TensorRT exports due to compatibility constraints. For reference, you can inspect the relevant code [here](https://github.com/ultralytics/ultralytics/blob/f932a611abf0ed0e359deecbb82fa8d361e23cfe/ultralytics/engine/exporter.py#L101-L121). If you require NMS for ONNX, a post-processing pipeline can be implemented externally as a workaround. Let us know if you need further clarification!

**Comment by rajhlinux** [comment]
> Thank you for your detailed follow-up! You are correct that currently, `nms=True` is not supported for ONNX export in the latest version (8.3.56) of Ultralytics. While `nms` functionality is available for certain formats, it is explicitly disabled for ONNX and TensorRT exports due to compatibility constraints. For reference, you can inspect the relevant code [here](https://github.com/ultralytics/ultralytics/blob/f932a611abf0ed0e359deecbb82fa8d361e23cfe/ultralytics/engine/exporter.py#L101-L121). If you require NMS for ONNX, a post-processing pipeline can be implemented externally as a workaround. Let us know if you need further clarification!

Hello Glenn, thanks for the reply.

Is it possible clarification can be provided in how to implement `nms` with ONNX for YOLO11? 

Thanks.

**Comment by Y-T-G** [comment]
@rajhlinux NMS support with all the major export formats is being tested in this PR

https://github.com/ultralytics/ultralytics/pull/18484/checks

**Comment by visionNoob** [comment]
@Y-T-G Its cool

**Comment by UltralyticsAssistant** [comment]
The associated pull request [#18484](https://github.com/ultralytics/ultralytics/pull/18484) has been merged, and it includes a potential fix that may resolve this issue! 🎉  

### Relevant Key Changes from the PR:
This update introduces extended export functionalities and post-processing improvements for object detection models, addressing several limitations in the workflow. Notable changes include:
- **Non-Maximum Suppression (NMS) Integration**: Exported models (TorchScript, ONNX, OpenVINO, TensorRT, TensorFlow) now include embedded NMS, providing seamless post-processing capabilities.
- **Enhanced Post-Processing Logic**: Predictions for various tasks (detection, segmentation, pose, and OBB) have been refined, including additional attribute support, like `keypoints`, for improved handling of prediction results.
- **Flexibility in Export Configurations**: Users can now toggle NMS during the export process, enabling easier deployment customization.
- **Better Metadata Handling**: Updates for improved export compatibility across formats.
- 🆙 Version upgraded to `8.3.67` for tracking this new functionality.

These changes aim to streamline the export process while improving runtime efficiency and usability for deployment scenarios.  

### Contributors ❤️
This valuable update was made possible due to contributions from [Y-T-G](https://github.com/Y-T-G), [lifafang](https://github.com/lifafang), [boorad](https://github.com/boorad), [glenn-jocher](https://github.com/glenn-jocher), [Laughing-q](https://github.com/Laughing-q), and [inisis](https://github.com/inisis). A big thank you to everyone involved! 🙌  

### How to Test the Changes:
To verify if this resolves your issue:
- Install the latest changes directly from the `main` branch:
  ```bash
  pip install git+https://github.com/ultralytics/ultralytics.git@main
  ```
- Or you can wait for the next official release on PyPI if preferred.

Let us know if this resolves the issue or if you encounter any remaining problems. Your feedback is invaluable in ensuring the robustness of these updates.  

Thanks again for bringing up this issue! 🙏 If anything persists or additional adjustments are needed, feel free to share more details.
